src/test/java/app/coronawarn/verification/service/TanServiceTest.java

```diff
@@ -231,6 +241,87 @@ public void testTeleTANFormat() {
     assertThat(tanService.isTeleTanValid("29zAABCL4-")).isFalse();
   }
 
+  @Test
+  public void testRateLimitCheckForTeleTan() {
+    config.getTan().getTele().getRateLimiting().setCount(TELE_TAN_RATE_LIMIT_COUNT);
+    config.getTan().getTele().getRateLimiting().setSeconds(TELE_TAN_RATE_LIMIT_SECONDS);
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isTrue();
+
+    for (int i = 0; i < TELE_TAN_RATE_LIMIT_COUNT - 1; i++) tanService.generateVerificationTeleTan();
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isTrue();
+
+    tanService.generateVerificationTeleTan();
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isFalse();
+  }
+
+  @Test
+  public void testRateLimitShouldNotCountNonTeleTan() {
+    config.getTan().getTele().getRateLimiting().setCount(TELE_TAN_RATE_LIMIT_COUNT);
+    config.getTan().getTele().getRateLimiting().setSeconds(TELE_TAN_RATE_LIMIT_SECONDS);
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isTrue();
+
+    for (int i = 0; i < TELE_TAN_RATE_LIMIT_COUNT + 1; i++) tanService.generateVerificationTan(TEST_TAN_SOURCE_OF_TRUST);
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isTrue();
+  }
+
+  @Test
+  public void testRateLimitShouldNotCountTeleTansOlderThanDefinedTimeWindow() {
+    config.getTan().getTele().getRateLimiting().setCount(TELE_TAN_RATE_LIMIT_COUNT);
+    config.getTan().getTele().getRateLimiting().setSeconds(TELE_TAN_RATE_LIMIT_SECONDS);
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isTrue();
+
+    VerificationTan tan1 = tanService.generateVerificationTan("tan1", TanType.TELETAN, TEST_TELE_TAN_SOURCE_OF_TRUST);
+    VerificationTan tan2 = tanService.generateVerificationTan("tan2", TanType.TELETAN, TEST_TELE_TAN_SOURCE_OF_TRUST);
+    tan1.setCreatedAt(LocalDateTime.now().minusSeconds(TELE_TAN_RATE_LIMIT_SECONDS + 1));
+    tan2.setCreatedAt(LocalDateTime.now().minusSeconds(TELE_TAN_RATE_LIMIT_SECONDS + 1));
+    tanService.saveTan(tan1);
+    tanService.saveTan(tan2);
+
+    for (int i = 0; i < TELE_TAN_RATE_LIMIT_COUNT - 1; i++) tanService.generateVerificationTeleTan();
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isTrue();
+
+    tanService.generateVerificationTeleTan();
+
+    assertThat(tanService.isTeleTanRateLimitNotExceeded()).isFalse();
+  }
+
+  @Test
+  public void testLogMessagesForRateLimits() throws NoSuchFieldException, IllegalAccessException {
+    Logger loggerMock = Mockito.mock(Logger.class);
+
+    // Usage of Relection API to "inject" mock for SLF4J Logger
```

Comment:
Java reflection is bad practice. Here we should come up with something more elegant

Reply:
As we have discussed in our team I have removed this test case.